source/module_io/write_Vxc.hpp

```diff
@@ -0,0 +1,210 @@
+        const double beta = 0.0;
+#ifdef __MPI
+        const int i1 = 1;
+        pdgemm_(&transa, &transb, &nbasis, &nbands, &nbasis, &alpha, V, &i1, &i1, pv.desc, c, &i1, &i1, pv.desc_wfc, &beta, Vc.data(), &i1, &i1, pv.desc_wfc);
+#else
+        dgemm_(&transa, &transb, &nbasis, &nbands, &nbasis, &alpha, V, &nbasis, c, &nbasis, &beta, Vc.data(), &nbasis);
+#endif
+        std::vector<double> cVc(p2d.nloc, 0.0);
+        transa = 'T';
+#ifdef __MPI
+        pdgemm_(&transa, &transb, &nbands, &nbands, &nbasis, &alpha, c, &i1, &i1, pv.desc_wfc, Vc.data(), &i1, &i1, pv.desc_wfc, &beta, cVc.data(), &i1, &i1, p2d.desc);
+#else
+        dgemm_(&transa, &transb, &nbands, &nbands, &nbasis, &alpha, c, &nbasis, Vc.data(), &nbasis, &beta, cVc.data(), &nbasis);
+#endif
+        return cVc;
+    }
+
+    inline double get_real(const std::complex<double>& c) { return c.real(); }
+    inline double get_real(const double& d) { return d; }
+
+    template<typename T>
+    double all_band_energy(const int ik, const std::vector<T>& mat_mo, const Parallel_2D& p2d, const ModuleBase::matrix& wg)
+    {
+        double e = 0.0;
+        for (int i = 0;i < p2d.get_row_size();++i)
+            for (int j = 0;j < p2d.get_col_size();++j)
+                if (p2d.local2global_row(i) == p2d.local2global_col(j))
+                    e += get_real(mat_mo[j * p2d.get_row_size() + i]) * wg(ik, p2d.local2global_row(i));
+        Parallel_Reduce::reduce_all(e);
+        return e;
+    }
+
+    template <typename T>
+    void set_gint_pointer(LCAO_Hamilt& uhm, typename TGint<T>::type*& gint);
+
+    template <>
+    void set_gint_pointer<double>(LCAO_Hamilt& uhm, typename TGint<double>::type*& gint)
+    {
+        gint = &uhm.GG;
+    }
+    template <>
+    void set_gint_pointer<std::complex<double>>(LCAO_Hamilt& uhm, typename TGint<std::complex<double>>::type*& gint)
+    {
+        gint = &uhm.GK;
+    }
+
+
+    /// @brief  write the Vxc matrix in KS orbital representation, usefull for GW calculation
+    template <typename TK, typename TR>
+    void write_Vxc(int nspin, int nbasis, int drank, const psi::Psi<TK>& psi, const UnitCell& ucell, Structure_Factor& sf,
+        const ModulePW::PW_Basis& rho_basis, const ModulePW::PW_Basis& rhod_basis, const ModuleBase::matrix& vloc,
+        const Charge& chg, LCAO_Hamilt& uhm, LCAO_Matrix& lm, Local_Orbital_Charge& loc,
+        const K_Vectors& kv, const ModuleBase::matrix& wg, Grid_Driver& gd)
+    {
+        ModuleBase::TITLE("ModuleIO", "write_Vxc");
+        const Parallel_Orbitals* pv = lm.ParaV;
+        int nbands = wg.nc;
+        // 1. real-space xc potential
+        // ModuleBase::matrix vr_xc(nspin, chg.nrxx);
+        double etxc = 0.0;
+        double vtxc = 0.0;
+        // elecstate::PotXC* potxc(&rho_basis, &etxc, vtxc, nullptr);
+        // potxc.cal_v_eff(&chg, &ucell, vr_xc);
+        elecstate::Potential* potxc = new elecstate::Potential(&rhod_basis, &rho_basis, &ucell, &vloc, &sf, &etxc, &vtxc);
```

Comment:
Why don't you try to get elecstate::PotXC from existed elecstate::Potential, which has member vector<PotBase*>.
Then the function "cal_v_eff" can help you.

Reply:
Sorry, I found Veff<OperatorLCAO> rely on Potential , maybe if this operator can only rely on PotBase will make the code more flexible, I will consider this question.

Reply:
> Why don't you try to get elecstate::PotXC from existed elecstate::Potential, which has member vector<PotBase*>. Then the function "cal_v_eff" can help you.
`Veff::contributeHR()` depends on `Potential::get_effective_v` which is not implemented in `PotBase`.